core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -1336,10 +1321,26 @@ private void commitIndexWriter(IndexWriter writer, Translog translog, String syn
             }
 
             if (logger.isTraceEnabled()) {
-                logger.trace("committing writer with commit data [{}]", commitData);
+                logger.trace("committing writer with commit data (max_seq_no excluded) [{}]", commitData);
             }
 
-            indexWriter.setCommitData(commitData);
+            indexWriter.setLiveCommitData(() -> {
+                /**
+                 * The user data in the commitData map contains data that must be evaluated *before*
+                 * Lucene flushes segments, including the local and global checkpoints amongst other values.
+                 * The maximum sequence number is different - we never want the maximum sequence number to be
+                 * less than the last sequence number to go into a Lucene commit, otherwise we run the risk
+                 * of re-using a sequence number for two different documents when restoring from this commit
+                 * point and subsequently writing new documents to the index.  Since we only know which Lucene
+                 * documents made it into the final commit after the {@link IndexWriter#commit()} call flushes
+                 * all documents, we defer computation of the max_seq_no to the time of invocation of the commit
+                 * data iterator (which occurs after all documents have been flushed to Lucene).
+                 */
+                final Map<String, String> deferredCommitData = new HashMap<>(commitData.size() + 1);
```

Comment:
I think it will be simpler to just capture the local and global checkpoints before we start and build one map here. This way we don't need two maps.